src/compiler/checker.ts

```diff
@@ -27515,73 +27515,433 @@ namespace ts {
-                    else if (inDestructuringPattern) {
-                        // Given the following situation:
-                        //    var c: {};
-                        //    [...c] = ["", 0];
-                        //
-                        // c is represented in the tree as a spread element in an array literal.
-                        // But c really functions as a rest element, and its purpose is to provide
-                        // a contextual type for the right hand side of the assignment. Therefore,
-                        // instead of calling checkExpression on "...c", which will give an error
-                        // if c is not iterable/array-like, we need to act as if we are trying to
-                        // get the contextual element type from it. So we do something similar to
-                        // getContextualTypeForElementExpression, which will crucially not error
-                        // if there is no index type / iterated type.
-                        const restElementType = getIndexTypeOfType(spreadType, numberType) ||
-                            getIteratedTypeOrElementType(IterationUse.Destructuring, spreadType, undefinedType, /*errorNode*/ undefined, /*checkAssignability*/ false) ||
-                            unknownType;
-                        elementTypes.push(restElementType);
-                        elementFlags.push(ElementFlags.Rest);
+                    else if (ctx===undefined || cidx === undefined){
+                        const type = checkExpressionForMutableLocation(e, checkMode, /* contextualType */ undefined, forceTuple);
+                        const flags = ElementFlags.Required;
+                        return { type, flags, assignable: false }; // there is no corresponding context type so it must be false.
+                    }
+                    else /* if (ctx!==undefined && cidx === undefined) */ {
+                        const cTarget = ctx.target;
+                        let cArgTypes = getTypeArguments(ctx);
+                        /**
+                         * NOTE!!!
+                         * The condition (cTarget.elementFlags.length < getTypeArguments(ctx).length) does occur.
+                         * It happens when there is tuple type which an unknow generic type in, e.g., [...T]
+                         * followed by an instantiation of that type, e.g. [...T] = ["hello"].
+                         * Then the types contained are [string, T] (length 2) but ctx.target.elementFlags is [ElementFlags.Required] (length 1).
+                         * The workaround here is to slice getTypeArguments(ctx)
+                         * to get an array of types which is the same length as the ctx.target.elementFlags.
+                         * This conformance in length is required when reading from the ends of the arrays.
+                         */
+                        if (cTarget.elementFlags.length<cArgTypes.length){

```

Comment:
The condition `(cTarget.elementFlags.length < getTypeArguments(ctx).length)` does occur. It happens when there is tuple type which an unknow generic type in, e.g., `[...T]` followed by an instantiation of that type, e.g. `[...T] = ["hello"]`. Then the types contained in `getTypeArguments(ctx)` are `[string, T]` (length 2) but `ctx.target.elementFlags` is `[ElementFlags.Required]` (length 1).

The workaround here is to slice `getTypeArguments(ctx)` to get an array of types which is the same length as the `ctx.target.elementFlags`.

This conformance in length is required when reading from the ends of the arrays.